# Eksperymenty

In [1]:
from BO2_projekt import Order, Car, select, max_aim, genetic, tabu

Utworzenie danych testowych - samochód o wymiarach 2.1 x 2.2 x 4.2, co daje pojemność równą 19.404 oraz 20 zamówień o zróżnicowanych objętościach, odległościach od naszej bazy oraz kierunkach.

In [2]:
car1 = Car(2.1, 2.2, 4.2, 11)

order1 = Order(8, 300, "n")
order2 = Order(8, 500, "e")
order4 = Order(1, 100, 'w')
order3 = Order(12, 800, "n")
order5 = Order(2, 200, "n")
order6 = Order(4, 100, "e")
order7 = Order(2, 200, 'e')
order8 = Order(20, 260, 'e')
order9 = Order(2, 200, 'e')
order10 = Order(7, 20, 'e')
order11 = Order(2, 600, 'w')
order12 = Order(12, 400, 'e')
order13 = Order(1, 900, 's')
order14 = Order(2, 280, 'n')
order15 = Order(9, 20, 'e')
order16 = Order(1, 800, 'w')
order17 = Order(17, 650, 'e')
order18 = Order(13, 400, 's')
order19 = Order(21, 30, 'n')
order20 = Order(17, 250, 'n')

order_list = [order1, order2, order3, order4, order5, order6, order7, order8, order9, order10, order11, order12, order13, order14, order15, order16, order17, order18, order19, order20]


Znalezienie najlepszego rozwiązania metodą dokładną (przeglądu wszystkich możliwych rozwiązań):

In [13]:
orders = select(order_list, car1)
print("\nMetoda 'brute force' - najlepsze możliwe rozwiązanie:", round(max_aim(orders, car1), 2))



Metoda 'brute force' - najlepsze mozliwe rozwiazanie: 33750.79


### Testowanie działania **algorytmu genetycznego**  

Dla każdego testu manipulujemy jednym parametrem, pozostałe pozostawiając jako defeaultowe, czyli dla przypomnienia:
- *kids_num* - liczba dzieci = 6
- *parents_num* - liczba rodziców (czyli ile osobników przetrwa selekcję i zostaną przekazane do krzyżowania) = 6
- *parents_mutants_num* - liczba rodziców przekazywanych do mutacji = 3
- *kids_mutants_num* - liczba dzieci przekazywanych do mutacji = 2
- *best_parents_num* - liczba najlepszych osobników z poprzedniej popalacji przekazywanych do kolejnej populacji = 4
- *way_of_crossing* - sposób krzyżowania = *hal-half* - dziecko to pierwsza połowa 'mamy' i druga połowa 'taty'  

pozostałe opcje krzyżowania to:
+ *start-middle-end* - dziecko to początek mamy, środek taty oraz końcówka mamy
+ *random* - punkt "przecięcia" rodziców jest wybierany losowo
+ *every-other* - co drugi element bierzemy z mamy, a co drugi z taty.



#### Kod generujący dane uzyskane z testów  
Dla każdego parametru algorytm był uruchamiany 10-krotnie, a wyniki zapisywane w macierzach.

In [4]:
cros = ['half-half', 'random', 'start-middle-end', 'every-other']
crossing_aims = []
crossing_iters = []
for way in cros:
    crossing_aim = []
    crossing_iter = []
    for _ in range(10):
        aim, iter = genetic(order_list, car1, 1000, way_of_crossing=way)
        crossing_aim.append(aim)
        crossing_iter.append(iter)

    crossing_aims.append(crossing_aim)
    crossing_iters.append(crossing_iter)

In [5]:
kids = [2, 5, 10]
kids_aims = []
kids_iters = []
for kid in kids:
    kids_aim = []
    kids_iter = []
    for _ in range(10):
        aim, iter = genetic(order_list, car1, 1000, kids_num=kid)
        kids_aim.append(aim)
        kids_iter.append(iter)

    kids_aims.append(kids_aim)
    kids_iters.append(kids_iter)

In [6]:
parents = [4, 8, 12]
parents_aims = []
parents_iters = []
for parent in parents:
    parents_aim = []
    parents_iter = []
    for _ in range(10):
        aim, iter = genetic(order_list, car1, 1000, parents_num=parent)
        parents_aim.append(aim)
        parents_iter.append(iter)

    parents_aims.append(parents_aim)
    parents_iters.append(parents_iter)

In [7]:
muts = [2, 6, 12]
muts_aims = []
muts_iters = []
for mut in muts:
    muts_aim = []
    muts_iter = []
    for _ in range(10):
        aim, iter = genetic(order_list, car1, 1000, kids_mutants_num=mut, parents_mutants_num=mut)
        muts_aim.append(aim)
        muts_iter.append(iter)

    muts_aims.append(muts_aim)
    muts_iters.append(muts_iter)

In [61]:
iters = [5, 10, 20, 50, 80, 100, 150, 200, 300, 500, 1000]
i_aims = []
i_iters = []
for i in iters:
    i_aim = []
    i_iter = []
    for _ in range(10):
        aim, iter = genetic(order_list, car1, i_max=i)
        i_aim.append(aim)
        i_iter.append(iter)

    i_aims.append(i_aim)
    i_iters.append(i_iter)

In [62]:
new_crossing_aims = [[round(a, 2) for a in b] for b in crossing_aims]
new_kids_aims = [[round(a, 2) for a in b] for b in kids_aims]
new_parents_aims = [[round(a, 2) for a in b] for b in parents_aims]
new_muts_aims = [[round(a, 2) for a in b] for b in muts_aims]
new_i_aims = [[round(a, 2) for a in b] for b in i_aims]

### Prezentacja oraz omówienie danych pozyskanych z testów.

1. Sposób krzyżowania

In [43]:
for i in range(len(new_crossing_aims)):
    print(f'{cros[i]:16}  {new_crossing_aims[i]}')


half-half         [32768.14, 33750.79, 33750.79, 32768.14, 32768.14, 32768.14, 32768.14, 32768.14, 32768.14, 33750.79]
random            [32768.14, 32768.14, 33750.79, 32768.14, 26348.11, 32768.14, 33750.79, 32768.14, 32768.14, 32768.14]
start-middle-end  [32768.14, 33750.79, 33750.79, 32768.14, 33750.79, 32768.14, 33750.79, 32768.14, 33750.79, 33750.79]
every-other       [33750.79, 32768.14, 33750.79, 33750.79, 33750.79, 33750.79, 33750.79, 32768.14, 32768.14, 33750.79]


Powyższe wyniki przedstawiają wartość funkcji celu uzyskanej dla 10 uruchomień algorytmu dla każdej z wartości parametrów. Żaden ze sposobów krzyżowania nie doprowadził do drastycznej zmiany wyników - za każdym razem otrzymano wynik optymalny (33750.79) lub bliski optymalnemu (~32000). Warto jednak zauważyć, że podczas gdy dla dwóch pierwszych sposobów otrzymano jedynie trzy- oraz dwukrotnie optymalne rozwiązanie (a nawet dla sposobu *random* raz uzyskano rozwiązanie niezadowalające (~26000)), tak dwa pozostałe sposoby krzyżowania dały sześcio- oraz siedmiokrotnie najlepsze możliwe rozwiązanie.

In [44]:
for i in range(len(crossing_iters)):
    print(f'{cros[i]:16}  {crossing_iters[i]}')

half-half         [211, 14, 8, 172, 9, 3, 35, 69, 104, 16]
random            [193, 69, 29, 62, 0, 22, 7, 25, 94, 46]
start-middle-end  [38, 26, 10, 134, 3, 169, 19, 37, 23, 66]
every-other       [7, 38, 104, 7, 27, 7, 3, 33, 52, 4]


Sposób krzyżowania prawdopodobnie nie ma wpływu na numer iteracji, w której znaleziono rozwiązanie końcowe.

2. Liczba dzieci

In [46]:
for i in range(len(new_kids_aims)):
    print(f'liczba dzieci = {kids[i]:2}  {new_kids_aims[i]}')

liczba dzieci =  2  [32768.14, 32768.14, 32768.14, 33750.79, 32768.14, 32768.14, 32768.14, 32768.14, 33750.79, 32768.14]
liczba dzieci =  5  [32768.14, 33750.79, 33750.79, 32768.14, 32768.14, 33750.79, 33750.79, 33750.79, 32768.14, 32768.14]
liczba dzieci = 10  [32768.14, 33750.79, 32768.14, 33750.79, 32768.14, 32768.14, 32768.14, 33750.79, 32768.14, 32768.14]


Dla każdej z liczby dzieci uzyskaliśmy podobne wyniki.

In [47]:
for i in range(len(kids_iters)):
    print(f'liczba dzieci = {kids[i]:2}  {kids_iters[i]}')

liczba dzieci =  2  [177, 16, 275, 43, 58, 26, 61, 226, 154, 31]
liczba dzieci =  5  [8, 91, 54, 82, 16, 53, 0, 0, 52, 76]
liczba dzieci = 10  [32, 10, 19, 53, 159, 104, 30, 6, 58, 34]


Wraz ze wzostem liczby dzieci, malał numer iteracji, w której uzyskano rozwiązanie optymalne.

3. Liczba rodziców

In [50]:
for i in range(len(new_parents_aims)):
    print(f'liczba rodziców = {parents[i]:2}  {new_parents_aims[i]}')

liczba rodziców =  4  [32768.14, 32768.14, 32768.14, 33750.79, 32768.14, 32768.14, 33750.79, 32768.14, 32768.14, 32768.14]
liczba rodziców =  8  [33750.79, 32768.14, 33750.79, 32768.14, 32768.14, 33750.79, 33750.79, 32768.14, 32768.14, 33750.79]
liczba rodziców = 12  [33750.79, 32768.14, 32768.14, 32768.14, 32768.14, 33750.79, 33750.79, 33750.79, 33750.79, 33750.79]


Wraz ze wzrostem liczby rodziców, algorytm częściej docierał do rozwiązania optymalnego (kolejno 2, 5 oraz 6 razy).

In [51]:
for i in range(len(parents_iters)):
    print(f'liczba rodziców = {parents[i]:2}  {parents_iters[i]}')

liczba rodziców =  4  [70, 49, 32, 2, 49, 21, 2, 81, 36, 66]
liczba rodziców =  8  [2, 6, 41, 22, 10, 22, 24, 27, 151, 17]
liczba rodziców = 12  [70, 58, 13, 18, 52, 20, 15, 110, 10, 3]


Liczba rodziców zdaje się nie mieć dużego wpływu na numer iteracji, w której znaleziono rozwiązanie końcowe.

4. Liczba mutacji

In [52]:
for i in range(len(new_muts_aims)):
    print(f'liczba mutacji = {muts[i]:2}  {new_muts_aims[i]}')

liczba mutacji =  2  [33750.79, 33750.79, 32768.14, 32768.14, 32768.14, 33750.79, 33750.79, 33750.79, 33750.79, 32768.14]
liczba mutacji =  6  [33750.79, 33750.79, 33750.79, 32768.14, 33750.79, 32768.14, 33750.79, 32768.14, 32768.14, 32768.14]
liczba mutacji = 12  [33750.79, 32768.14, 33750.79, 33750.79, 33750.79, 33750.79, 32768.14, 33750.79, 33750.79, 32768.14]


Dla każdej liczby mutacji algorytm podobnie często znajdował rozwiązanie optymalne (opdowiednio 6, 5 oraz 7 razy)

In [53]:
for i in range(len(muts_iters)):
    print(f'liczba mutacji = {muts[i]:2}  {muts_iters[i]}')

liczba mutacji =  2  [22, 13, 83, 73, 7, 13, 39, 57, 21, 109]
liczba mutacji =  6  [10, 4, 2, 5, 2, 7, 9, 7, 7, 7]
liczba mutacji = 12  [11, 40, 5, 2, 11, 11, 8, 4, 3, 15]


Wraz ze wzrostem liczby mutacji, numer iteracji, w której znaleziono ostateczne rozwiązanie wydaje się nieznacznie maleć.

5. Liczba iteracji

In [63]:
for i in range(len(new_i_aims)):
    print(f'liczba iteracji = {iters[i]:4}  {new_i_aims[i]}')

liczba iteracji =    5  [26348.11, 28914.76, 29304.93, 33750.79, 29286.7, 28175.42, 26909.65, 30813.25, 28175.42, 31984.79]
liczba iteracji =   10  [28235.37, 32768.14, 27382.64, 33750.79, 28242.91, 28242.91, 33750.79, 26336.47, 31912.03, 33750.79]
liczba iteracji =   20  [30048.54, 28851.42, 31984.79, 33750.79, 32768.14, 33750.79, 33750.79, 28851.42, 28242.91, 32768.14]
liczba iteracji =   50  [32768.14, 30813.25, 32768.14, 33750.79, 30813.25, 28851.42, 33750.79, 33750.79, 32768.14, 32768.14]
liczba iteracji =   80  [31984.79, 32768.14, 28242.91, 30047.05, 33750.79, 30813.25, 32768.14, 31912.03, 32768.14, 32768.14]
liczba iteracji =  100  [32768.14, 32768.14, 30813.25, 33750.79, 33750.79, 31984.79, 32768.14, 33750.79, 30813.25, 31984.79]
liczba iteracji =  150  [33750.79, 33750.79, 32768.14, 33750.79, 32768.14, 33750.79, 33750.79, 33750.79, 32768.14, 30813.25]
liczba iteracji =  200  [32768.14, 33750.79, 32768.14, 28914.76, 33750.79, 32768.14, 33750.79, 30813.25, 33750.79, 33750.79]
l

Algorytm za każdym razem znajduje rozwiązanie optymalne (33750.79) lub jemu bliskie (32768.14) już od 150 iteracji. Warto również zauważyć, że już dla ograniczenia do 10 iteracji raz udało się znaleźć rozwiązanie optymalne, a wyniki dla ograniczenia 5 iteracji również są całkiem dobre.

In [64]:
for i in range(len(i_iters)):
    print(f'liczba mutacji = {iters[i]:4}  {i_iters[i]}')

liczba mutacji =    5  [0, 0, 0, 0, 0, 4, 2, 4, 0, 2]
liczba mutacji =   10  [5, 6, 9, 2, 1, 3, 7, 6, 4, 2]
liczba mutacji =   20  [6, 2, 15, 8, 9, 4, 17, 5, 0, 9]
liczba mutacji =   50  [43, 18, 1, 10, 40, 10, 49, 24, 22, 41]
liczba mutacji =   80  [10, 54, 8, 78, 59, 10, 55, 77, 10, 48]
liczba mutacji =  100  [75, 38, 16, 27, 2, 9, 7, 16, 46, 24]
liczba mutacji =  150  [122, 6, 0, 17, 32, 33, 50, 7, 17, 6]
liczba mutacji =  200  [78, 41, 24, 2, 60, 5, 78, 22, 141, 28]
liczba mutacji =  300  [144, 14, 17, 157, 19, 23, 28, 121, 253, 14]
liczba mutacji =  500  [13, 97, 64, 123, 39, 6, 12, 140, 209, 67]
liczba mutacji = 1000  [123, 12, 12, 6, 96, 108, 68, 8, 178, 253]


Analizując numer iteracji, w której znaleziono rozwiązanie końcowe, widać, że jedynie w niewielu przypadkach ta liczba przekracza 150 (gdy ma taką możliwość), natomiast maksymalnym wynikiem jest granica około 260 iteracji.

### Testowanie działania **algorytmu tabu search**  

In [66]:
i_t = [5, 10, 20, 50, 100, 150, 300, 500, 1000, 2000, 3000, 4000, 5000]
tabu_aims = []
tabu_iters = []
for it in i_t:
    tabu_aim = []
    tabu_iter = []
    for _ in range(5):
        aim, iter = tabu(order_list, car1, it)
        tabu_aim.append(aim)
        tabu_iter.append(iter)
    tabu_aims.append(tabu_aim)
    tabu_iters.append(tabu_iter)

In [67]:
new_tabu_aims = [[round(a, 2) for a in b] for b in tabu_aims]

In [69]:
for i in range(len(new_tabu_aims)):
    print(f'liczba iteracji = {i_t[i]:4}  {new_tabu_aims[i]}')

liczba iteracji =    5  [24161.53, 26720.22, 23319.21, 23347.01, 15313.44]
liczba iteracji =   10  [19481.83, 19985.73, 18426.19, 25761.65, 12333.55]
liczba iteracji =   20  [15035.72, 18967.21, 26019.33, 24484.66, 18717.12]
liczba iteracji =   50  [26399.81, 31912.03, 27492.2, 27382.64, 19117.77]
liczba iteracji =  100  [31956.25, 27263.73, 31956.25, 29286.7, 28544.58]
liczba iteracji =  150  [28152.15, 33750.79, 26263.71, 29834.07, 27486.85]
liczba iteracji =  300  [32768.14, 30978.17, 30048.54, 33750.79, 32768.14]
liczba iteracji =  500  [31956.25, 33750.79, 26336.47, 30201.27, 33750.79]
liczba iteracji = 1000  [33750.79, 31956.25, 33750.79, 31912.03, 31912.03]
liczba iteracji = 2000  [33750.79, 31912.03, 31956.25, 33750.79, 33750.79]
liczba iteracji = 3000  [33750.79, 33750.79, 33750.79, 33750.79, 33750.79]
liczba iteracji = 4000  [33750.79, 33750.79, 33750.79, 33750.79, 33750.79]


Na powyższych wynikach wyraźnie widać, że wraz ze wzrostem limitu iteracji, wynik ulega poprawie. Algorytm od 3000 iteracji za każdym razem dociera do rozwiązania optymalnego, natomiast po raz pierwszy znalazł je już dla 150 iteracji.

In [82]:
lengths = [2, 3, 5, 10, 20, 30, 50, 100, 200, 300, 400, 500]
length_aims = []
length_iters = []
for length in lengths:
    length_aim = []
    length_iter = []
    for _ in range(5):
        aim, iter = tabu(order_list, car1, i_max=3000, tabu_length=length)
        length_aim.append(aim)
        length_iter.append(iter)
    length_aims.append(length_aim)
    length_iters.append(length_iter)

In [84]:
new_length_aims = [[round(a, 2) for a in b] for b in length_aims]
for i in range(len(new_length_aims)):
    print(f'rozmiar tabu = {lengths[i]:4}  {new_length_aims[i]}')

rozmiar tabu =    2  [32768.14, 33750.79, 33750.79, 33750.79, 33750.79]
rozmiar tabu =    3  [33750.79, 33750.79, 33750.79, 32986.08, 33750.79]
rozmiar tabu =    5  [33750.79, 33750.79, 33750.79, 31984.79, 33750.79]
rozmiar tabu =   10  [33750.79, 33750.79, 33750.79, 33750.79, 33750.79]
rozmiar tabu =   20  [33750.79, 33750.79, 33750.79, 33750.79, 33750.79]
rozmiar tabu =   30  [33750.79, 33750.79, 33750.79, 33750.79, 33750.79]
rozmiar tabu =   50  [33750.79, 33750.79, 33750.79, 33750.79, 33750.79]
rozmiar tabu =  100  [33750.79, 33750.79, 33750.79, 33750.79, 33750.79]
rozmiar tabu =  200  [33750.79, 33750.79, 33750.79, 33750.79, 33750.79]
rozmiar tabu =  300  [33750.79, 33750.79, 33750.79, 33750.79, 32768.14]
rozmiar tabu =  400  [33750.79, 33750.79, 33750.79, 33750.79, 33750.79]
rozmiar tabu =  500  [33750.79, 33750.79, 33750.79, 33750.79, 33750.79]


Jedyny 'gorszy' wynik trafił się dla długości listy tabu równej 2, natomiast dalsze modyfikacje rozmiaru listy nie miały wpływu na wynik końcowy przy liczbie iteracji równej 3000.

In [85]:
for i in range(len(new_tabu_aims)):
    print(f'liczba iteracji = {lengths[i]:4}  {length_iters[i]}')

liczba iteracji =    2  [2727, 253, 2599, 996, 0]
liczba iteracji =    3  [929, 558, 160, 648, 167]
liczba iteracji =    5  [48, 129, 1097, 702, 638]
liczba iteracji =   10  [1321, 1234, 1192, 852, 380]
liczba iteracji =   20  [1483, 2087, 1765, 2347, 1524]
liczba iteracji =   30  [2140, 1092, 1406, 152, 2165]
liczba iteracji =   50  [304, 41, 307, 1835, 523]
liczba iteracji =  100  [1869, 157, 1018, 1079, 9]
liczba iteracji =  200  [140, 1234, 70, 2664, 1149]
liczba iteracji =  300  [97, 226, 510, 2947, 2388]
liczba iteracji =  400  [1445, 381, 216, 117, 887]
liczba iteracji =  500  [270, 2004, 325, 2451, 605]


Rozmiar listy tabu wydaje się nie mieć wpływu na iterację, w której znaleziono ostateczny wynik.

## Prezentacja wyników wszystkich trzech algorytmów

In [87]:
    result_brute, aim_brute = max_aim(orders, car1, full_return=True)
    print("\nNajlepsza możliwa wartość funkcji celu:       ", aim_brute)
    print(result_brute)
    result_tabu, aim_tabu, iter_tabu = tabu(order_list, car1, 10000, full_return=True)
    result_genetic, aim_genetic, iter_genetic = genetic(order_list, car1, 300, full_return=True)
    print("\nAlgorytm tabu       ===> wartość funkcji celu:", aim_tabu)
    print(result_tabu)
    print("\nAlgorytm genetyczny ===> wartość funkcji celu:", aim_genetic)
    print(result_genetic)


Najlepsza możliwa wartość funkcji celu:        33750.7919390138
[Vol = 8, Dist = 500, Dir = e;, Vol = 1, Dist = 100, Dir = w;, Vol = 2, Dist = 200, Dir = n;, Vol = 2, Dist = 200, Dir = e;, Vol = 2, Dist = 600, Dir = w;, Vol = 1, Dist = 900, Dir = s;, Vol = 2, Dist = 280, Dir = n;, Vol = 1, Dist = 800, Dir = w;]

Algorytm tabu       ===> wartość funkcji celu: 33750.7919390138
[Vol = 8, Dist = 500, Dir = e;, Vol = 1, Dist = 100, Dir = w;, Vol = 2, Dist = 200, Dir = e;, Vol = 2, Dist = 200, Dir = e;, Vol = 2, Dist = 600, Dir = w;, Vol = 1, Dist = 900, Dir = s;, Vol = 2, Dist = 280, Dir = n;, Vol = 1, Dist = 800, Dir = w;]

Algorytm genetyczny ===> wartość funkcji celu: 33750.7919390138
[Vol = 8, Dist = 500, Dir = e;, Vol = 1, Dist = 100, Dir = w;, Vol = 2, Dist = 200, Dir = n;, Vol = 2, Dist = 200, Dir = e;, Vol = 2, Dist = 600, Dir = w;, Vol = 1, Dist = 900, Dir = s;, Vol = 2, Dist = 280, Dir = n;, Vol = 1, Dist = 800, Dir = w;]
